### ISSUE TITLE
Ideas for new developers

### ISSUE THREAD

**Issue by Zarel** [Zarel/Pokemon-Showdown#2444]
This is an issue of suggestions for things people who want to start contributing to Pokémon Showdown can work on.

First, the Development room is the right place to discuss working on new features: http://psim.us/dev

- **Fix mechanics bugs** – #2367 tracks the mechanics bugs we're aware of.

- **Fixes to text hiding** – Text hiding in `/roomban`, `/hidetext`, etc currently work by checking for user objects, which expire after 40 minutes of being offline. There are several possible improvements here:
  1. Actually clear the text from the roomlog, so users joining after the text is hidden don't see it at all. This is useful if the text being hidden is a secret, and also saves bandwidth and room clutter.
  2. To support very inactive rooms, scan the roomlog for the passed username, as well as checking for user objects.
  3. To support users who change usernames multiple times, scan the roomlog for the user's entire `prevNames` list.

- **Move timer code from rooms.js to simulator.js** – A pretty straightforward refactor.

- **Rewrite sockets.js in Go** – You'll probably need https://github.com/igm/sockjs-go – this is a pretty hard task but doesn't need anyone particularly familiar with the codebase. Node.js's support for WebSocket seems to have multiple bugs which are very hard to fix (see nodejs/node#3072 and nodejs/node#5469) so changing languages may be easier. Fortunately, PS's architecture is multi-process which makes it possible to change language for one process without having to rewrite the entire thing from scratch.

- **Refactor tournaments to use RoomGamePlayer** – See #2331 – A self-contained refactor.

**Comment by robotmayo** [comment]
Whats the current status on the Golang rewrite? I'm a Javascript dev looking to get into Golang. If possible id like to assist whoever is leading that.

**Comment by Zarel** [comment]
@robotmayo No one's started on that yet. Feel free to try it yourself. In addition to golang, you'll need to be able to do IPC between Node and Go, which shouldn't be too difficult. You may be interested in https://github.com/jgranstrom/gonode

**Comment by Joimer** [comment]
@robotmayo Maybe @xfix Would be interested in that (sorry if you aren't, xfix :)). If you start the project, I'd like to follow it, as I'd like to get more into Golang as well.

**Comment by Slayer95** [comment]
@Zarel, I see the list randomly growing with no notification.
Each item should probably be its own issue.

**Comment by Zarel** [comment]
I think that's unnecessary, but I guess I can comment every time I add something.

**Comment by acz13** [comment]
quick folder export implementation at Zarel/Pokemon-Showdown-Client#756

**Comment by Morfent** [comment]
I'm working on the sockets.js Golang rewrite.

**Comment by BlackiDucki** [comment]
Is "Give battles a user list like chatrooms" done?

**Comment by Zarel** [comment]
Nope. iirc Bumba was investigating it but no code's been written yet?

**Comment by bumbadadabum** [comment]
I gave up after a while, was thinking about picking it up again but if
anyone else wants to give it a whirl feel free.

**Comment by CrestfallPS** [comment]
panpawn made it so /userlist displays those in a battle room but not an actual list (which would look bad anyways imo)

**Comment by bumbadadabum** [comment]
/userlist is a temporary solution, we still want the actual userlist.

**Comment by panpawn** [comment]
^

**Comment by acz13** [comment]
>Teambuilder import folder – A way to import a single folder from the teambuilder.

can't we just Ctrl-V?

**Comment by Morfent** [comment]
The sockets.go rewrite description should probably be updated:

>Fortunately, PS's architecture is multi-process which makes it possible to change language for one process without having to rewrite the entire thing from scratch. We'll need to figure out exactly how child_process's IPC works, though.

It uses TCP connections by default. On the Node end, some manual buffering is needed for messages received upstream from the child process because of an internal buffer cap for incoming messages in `net.Connection`. `cluster` IPC messages use objects with a message type and data values, but that's not an option for IPC with Go child processes since Node and Go cannot parse objects/structs sent by each other. Instead they stringify/parse messages as JSON and delimit them with characters that are escaped to hexadecimal/unicode escape codes as JSON to avoid exploits with message parsing.

**Comment by Maxalexander-pi** [comment]
Is anyone working on the userlist for battles? I can give that a try if no one is already

**Comment by panpawn** [comment]
>Is anyone working on the userlist for battles? I can give that a try if no one is already

Pretty sure no one is working on that right now, feel free! It should be minimized by default like the userlist in chat rooms are on mobile.

**Comment by Zarel** [comment]
Also now is probably a bad time for client changes; we have a pending React rewrite for the client...

**Comment by bgottlob** [comment]
Is anyone working on refactoring the old tests to use `makeChoices`? I'd like to try that if not.

**Comment by Zarel** [comment]
Not currently; please feel free to do that.

**Comment by chichi148** [comment]
Anyone working on the advanced search for teambuilding? If not I would like to give it a shot.

**Comment by HalmonLui** [comment]
Anyone working on the import of a teambuilding folder? I'd like to attempt it nobody is.

**Comment by littlefoot22** [comment]
took a shot at **Advanced search in Teambuilder**

https://github.com/Zarel/Pokemon-Showdown-Client/pull/1133

**Comment by HoeenCoder** [comment]
Added:

**Support other mods in commands such as /weak** – Other examples include /coverage, /effectiveness, and /statcalc. Should mostly be as simple as supporting an optional argument (/weak pokemon/types, mod). When used in a battle they should default to the format's mod, otherwise default to the current generation. See /data for a current implementation.

**Comment by Zarel** [comment]
- ~~**Migrate tests to Jest**~~ – Jest (https://jestjs.io/) supports parallel test-running, which would be nice to have over Mocha. Most of its API surface is similar, so this shouldn't take too long.

Removed: Jest turned out to be much slower

- ~~**Rewrite sockets.js in Go**~~ (Morfent is working on this) – You'll probably need https://github.com/igm/sockjs-go – this is a pretty hard task but doesn't need anyone particularly familiar with the codebase. Node.js's support for WebSocket seems to have multiple bugs which are very hard to fix (see nodejs/node#3072 and nodejs/node#5469) so changing languages may be easier. Fortunately, PS's architecture is multi-process which makes it possible to change language for one process without having to rewrite the entire thing from scratch. We'll need to figure out exactly how `child_process`'s IPC works, though.

Removed: SockJS no longer has the perf issues that were plaguing us when I suggested this, making this migration overall no longer worth it.

**Comment by scheibo** [comment]
- **Migrate additional files to Typescript** – The files which are already written in [typed-JS](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html) are likely to be easier. #5210 and #5217 provide some examples of migrations, though it should be possible to convert on a file-by-file basis instead of converting whole directories. 
- [**Popular suggestions from Smogon**](https://www.smogon.com/forums/threads/suggestions-for-pok%C3%A9mon-showdown.3534365/) – Not necessarily up to date, but "More common Suggestions" lists a host of user requested features (as does the rest of the thread), several which would be fairly straightforward to implement for a new developer. Reach out on http://psim.us/dev before starting work on any of them to determine if the suggested feature is actually a good fit.

**Comment by asingh04** [comment]
Hi, I am a TypeScript developer, and I noticed the point `Migrate additional files to Typescript`. Is there any issue related to Typescript i could work on?

**Comment by Zarel** [comment]
@asingh04 yes, you can just migrate files in `server/` from `.js` to `.ts`. It'll require some screwing around; feel free to ask for help if you need it.

**Comment by asingh04** [comment]
@Zarel do i need to create an issue/ticket on GitHub first, or can i start working on it and raise the PR?

**Comment by asingh04** [comment]
I have made a ticket as [#5528](https://github.com/Zarel/Pokemon-Showdown/issues/2444) for migration of a file

**Comment by Zarel** [comment]
- ~~**Advanced search in Teambuilder**~~ – Some users want to be able to filter teams by what pokemon they have. Just searching for IDs in the packed team should be enough to support filtering by pokemon, ability, move, and item. WARNING: if you work on this, work on this in the Preact rewrite – https://github.com/Zarel/Pokemon-Showdown-Client/pull/1156

Removed: Done thanks to @littlefoot22 in https://github.com/Zarel/Pokemon-Showdown-Client/pull/1133

**Comment by TheJetOU** [comment]
> Support renaming chatrooms – Now that we have a |rename| message type, the server-side is relatively straightforward: make sure there's no roomgame running, move the entry in Rooms.rooms, move the entry in every user.inRooms, move the entry in every connection.inRooms, move the modlog, move the roomlog, check polls, and I think that's it? Test to make sure.

This is done in https://github.com/smogon/pokemon-showdown/commit/34e215c7a617347ce088d5e100114840df9064a6 and https://github.com/smogon/pokemon-showdown-client/commit/e7c645647b60cd6d9a7447b7ee7d2bcf80ba6945.

**Comment by scheibo** [comment]
> This is done in [34e215c](https://github.com/smogon/pokemon-showdown/commit/34e215c7a617347ce088d5e100114840df9064a6) and [smogon/pokemon-showdown-client@e7c6456](https://github.com/smogon/pokemon-showdown-client/commit/e7c645647b60cd6d9a7447b7ee7d2bcf80ba6945).

Removed, thanks.

**Comment by spoorthyv** [comment]
Hey everyone, Im a UI/UX Designer looking to contribute. Is there any interest in doing a full UI redesign? If so, how should I go about doing that?

**Comment by Zarel** [comment]
> Hey everyone, Im a UI/UX Designer looking to contribute. Is there any interest in doing a full UI redesign? If so, how should I go about doing that?

Join us on Discord! https://spo.ink/dev

We'd probably want to start by seeing mockups/sketches.

**Comment by arcreaco** [comment]
You Have to Work on CSS of PS!

**Comment by lkyhfx** [comment]
Hi, I notice that the showdown website only support English, and there are many lovers around the world that do not familiar with English. I am from China, and I know lots of people in China have to use translation tool to enjoy the battles. Do add other language support possible? I have an idea that can add Chinese support easily (also works for other language I think).  `wiki.52poke.com` is a Pokémon wiki website, it contains lots of information both in Chinese and English (such as Pokémons, items, moves). We can scrap the website then 'translate' the showdown website into Chinese. I think only some scripts is needed to do this job. Although the result  isn't perfect, the experience will be improved greatly.

**Comment by alexandmi** [comment]
Hello guys, i would like to make an SRS document for the website and its software. Is there any previous kind of documentation of this sort? I don't know if i have to post it here or the dev chat since it's not about developing or fixing bugs.

**Comment by gigalh128** [comment]
> [Bug reports from Smogon](https://www.smogon.com/forums/threads/bug-reports-v3-read-original-post-before-posting.3634749/) – Not necessarily all valid (be sure to confirm with a researcher or on the cartridge whether any reported mechanics issues are actually issues), but often the first place recent regressions are reported.

The link here is outdated

**Comment by HoeenCoder** [comment]
Bug reports have been moved to their own forum located here: https://www.smogon.com/forums/forums/ps-bug-reports.876/ It should be easier to see what still needs work and what doesn't.

**Comment by KingDoomYTX** [comment]
weird question, do i have to be unlocked to be a dev, or can i be locked and still be a dev, or do I have to be staff?
